File: day2/day2.py
import sys

def checksum(lines, checksum_func=sum, separator="\t"):
    """
    Calculates the checksum of a set of lines in a spreadsheet.

    Parameters:
        lines (list<str>): The lines in the spreadsheet.
        checksum_func (func, optional): The function to calculate the checksum
                                        for each line in the spreadsheet.
                                        Default: sums all numbers.
        separator (str, optional): The character the separates numbers in
                                   the spreadsheet.
    """
    checksum = 0
    for line in lines:
        numbers = [int(num) for num in line.split(separator)]
        checksum += checksum_func(numbers)
    return checksum

def even_numbers(numbers):
    """
    Find the division of the two numbers in the numbers list that are
    evenly divisible.

    Parameters:
        numbers (list<int>): The list of numbers.

    Returns:
        (int): The division of the two evenly divisible numbers.
    """
    # TODO: Better solution?
    for num in numbers:
        for num1 in numbers:
            if num % num1 == 0 and num != num1:
                return num // num1

def run(stdin):
    """
    Takes problem input and yields solutions.

    Parameters:
        stdin (str): The input to the problem as a string.

    Yields:
        (*): The solutions to the problem.
    """
    yield checksum(stdin.splitlines(),
                   checksum_func=lambda nums: max(nums) - min(nums))
    yield checksum(stdin.splitlines(), checksum_func=even_numbers)

if __name__ == "__main__":
    if __name__ == "__main__":
        results = run(sys.stdin.read())
        for result in results:
            print(result)

File: day2/day2.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <limits.h>

#include "../api.h"

int checksum(char** lines, int lineCount, int (*checksum_func)(int*, int)) {
    int checksum = 0;
    for (int i = 0; i < lineCount; i++) {
		SplitResult* result = split(lines[i], '\t');

		int* numbers = malloc(sizeof(int) * result->count);

		for (int j = 0; j < result->count; j++) {
			sscanf(result->strings[j], "%d", &(numbers[j]));
		}

		checksum += checksum_func(numbers, result->count);
    }
    return checksum;
}

int max_minus_min(int* numbers, int numberCount) {
	int max = 0;
	int min = INT_MAX;

	for (int i = 0; i < numberCount; i++) {
		if (numbers[i] > max) {
			max = numbers[i];
		}
		if (numbers[i] < min) {
			min = numbers[i];
		}
	}

	return max - min;
}

int even_numbers(int* numbers, int numberCount) {
	for (int i = 0; i < numberCount; i++) {
		for (int j = 0; j < numberCount; j++) {
			if ((numbers[i] % numbers[j] == 0) && (numbers[i] != numbers[j])) {
				return numbers[i] / numbers[j];
			}
		}
	}
	return 0;
}

int main(int argc, char** argv) {
	Input* input = load_input();

    printf("%d\n", checksum(input->lines, input->count, max_minus_min));
    printf("%d\n", checksum(input->lines, input->count, even_numbers));
}